### ISSUE TITLE
Docker Logs HTTP Format is Ambiguous

### ISSUE THREAD

**Issue by dperny** [moby/moby#31956]
`docker logs` (and by extension, `docker service logs`, which duplicates the `docker logs` api format) is ambiguous and cannot be reliably parsed by machines.

Specifically, there are two problems: 

<strike>First, there is no delimiter for the end of log lines, so log lines that include newlines cannot be parsed. You can't tell when one log line ends and another begins. The `stdcopy` marshaling format sort of fixes this by including a message length, but there is no such message length for TTY logs, which are sent raw. </strike>

**EDIT**: the above might not be a problem, actually, I think I'm confusing control codes (inserting `^M`, a literal enter key press) with newlines. Something is still weird with it, but I'm not sure what, and I think it's unrelated. 

Second, `details` of a log allow for unbounded user-entered data, which means that you can confuse any parser of logs by putting things like spaces, commas, and equals signs in the key or value field of your details.

For example, consider the following. In the first example, we can split out the details by taking the second space ` `. The spaces in the environment variable in the second part make that log impossible to parse. When do the details end and the log message begins? A user might fight it intuitive but a computer can never know without the context of the program being logged.

I have an outstanding PR #31952 that works around this internally, so that swarmkit does not have to parse log messages from the docker format. However, as it is currently design, swarmkit embeds context of the node, service, and task of a log message as details to send over the wire, and then parses them out for a better user interface. We can't continue to provide that nice UI *and* support user `details` if we don't fix this problem. 

```
$ docker run -d --log-opt env=asdf,jkl --name testdetails1 -e asdf=wombo -e jkl=combo flyinprogrammer/logtester
296aca12eff1836389c9d7ef4003d16fa13a302c597ce919680cd1b29628c22c

 $ docker logs --timestamps --details testdetails1
2017-03-20T22:41:37.438236633Z asdf=wombo,jkl=combo + exec app
2017-03-20T22:41:37.466396100Z asdf=wombo,jkl=combo 2017-03-20 22:41:37.46574108 +0000 UTC stderr msg: 1
2017-03-20T22:41:37.466479201Z asdf=wombo,jkl=combo 2017-03-20 22:41:37.46574108 +0000 UTC stdout msg: 1
2017-03-20T22:41:42.470094993Z asdf=wombo,jkl=combo 2017-03-20 22:41:42.469542351 +0000 UTC stdout msg: 2
2017-03-20T22:41:42.470306141Z asdf=wombo,jkl=combo 2017-03-20 22:41:42.469542351 +0000 UTC stderr msg: 2
2017-03-20T22:41:47.473416235Z asdf=wombo,jkl=combo 2017-03-20 22:41:47.473037886 +0000 UTC stderr msg: 3
2017-03-20T22:41:47.473535493Z asdf=wombo,jkl=combo 2017-03-20 22:41:47.473037886 +0000 UTC stdout msg: 3
2017-03-20T22:41:52.477177753Z asdf=wombo,jkl=combo 2017-03-20 22:41:52.476695927 +0000 UTC stdout msg: 4
2017-03-20T22:41:52.477305202Z asdf=wombo,jkl=combo 2017-03-20 22:41:52.476695927 +0000 UTC stderr msg: 4
```

```
$ docker run -d --log-opt env=asdf,jkl --name testdetails2 -e asdf="wombo combo" -e jkl="your logs broke" flyinprogrammer/logtester
e33776e91b8b6a46ab6c215922584b2c4836ccfc3238e02af323bdd127fa3414

$ docker logs --timestamps --details testdetails2
2017-03-20T22:43:40.085280619Z asdf=wombo combo,jkl=your logs broke + exec app
2017-03-20T22:43:40.087018949Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:40.086539819 +0000 UTC stdout msg: 1
2017-03-20T22:43:40.087171467Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:40.086539819 +0000 UTC stderr msg: 1
2017-03-20T22:43:45.090802840Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:45.090365361 +0000 UTC stderr msg: 2
2017-03-20T22:43:45.090883844Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:45.090365361 +0000 UTC stdout msg: 2
2017-03-20T22:43:50.094351612Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:50.093726656 +0000 UTC stdout msg: 3
2017-03-20T22:43:50.094567255Z asdf=wombo combo,jkl=your logs broke 2017-03-20 22:43:50.093726656 +0000 UTC stderr msg: 3
```

There are a couple of possible solutions for this. The first is a hack: do something like URL encode details in log messages when sending them over the wire so that things like spaces, commas, equals signs, and new lines cannot screw up log parsing, and then de-URL encode those things when we display things for the user on the CLI. This means we change the REST API somewhat (in a way that shouldn't affect any users, because nobody should be able to rely on the broken grammar that we're fixing) while leaving the CLI with identical behavior. 

The second is to switch to a different log message format that is more structured and which has an unambiguous grammar. Some people (@aluzzardi among them) have suggested returning logs in some standard like Syslog format, which could be more easily parsed. I have also heard talk of maybe having, like, a content-type header send down a different, more structured format (protobuf or similar). 

The third is to just, like, disallow spaces, commas, equals signs, and new lines in details. I don't think that's workable, though. 

I'm not sure what the right answer is, but the wrong answer, both internally to the project and for our users, is to stay the course with what is essentially a broken API.

PS: I've accidentally turned myself into a low-key domain expert on certain parts of logs over the past couple of months, which means I've probably got a bit of an expert's blind spot going on with the level of detail I've explained. Feel free to let me know if there's any part I need to explain in greater depth.

**Comment by dperny** [comment]
This is a semi-blocking issue for getting service logs to GA (#31399 as a meta-issue) because we cannot support details while the format is broken, so either the format has to be fixed or support for details has to be dropped.

**Comment by cpuguy83** [comment]
I don't think the log format was designed to do anything other than just be regurgitated by the API consumer and not really parsed except for stdout/stderr.

Definitely 👍  to improving this.
A simple change to the media type is probably not enough. All API clients < version N should be served the old format at the very least.

I like the idea of using a standard log format however the implementation on these has proven difficult, and even within the syslog format there are multiple RFCs for which we'd have to choose one: https://github.com/RackSec/srslog/blob/master/formatter.go

If we provide protobufs there is a well defined structure that anyone can generate a working client from which should be relatively efficient to stream as well.

Note that in #28403 we are working on a logger plugin implementation that uses protobufs as well. It's not exactly the same since the stream format only needs to handle the log stream type, the actual log data, and the timestamp and none of the extra details since these are initialized on driver startup.

**Comment by aluzzardi** [comment]
👍 overall

I think:
1) Short term we should "enhance" the current format with URL encoding, at least to make it somewhat machine parsable
2) Medium term, we should provide an additional machine-friendly format such as syslog. I think the client could negotiate the format (e.g. `?format=syslog`) so we don't break compat and eventually we deprecate the current format

I would avoid protobufs because, even though I'm a huge fan, it's really a hassle to work with if you're not running a proto shop already

**Comment by cpuguy83** [comment]
I generally consider formatting a job of the client. We're in a really bad situation now because formatting is on the daemon, but I suppose swapping an ill-conceived and ill-defined format for one that is well defined is a step forward... I'm afraid of adding more technical debt here, though.

**Comment by dperny** [comment]
fixed by #32996 cc @thaJeztah
